harness/determined/common/experimental/metrics.py

```diff
@@ -0,0 +1,90 @@
+import dataclasses
+import datetime
+from typing import Any, Dict, List, Optional, cast
+
+from determined.common import util
+from determined.common.api import bindings
+
+
+@dataclasses.dataclass
+class TrialMetrics:
+    """
+    Specifies a metric that the trial reported.
+
+    Attributes:
+        trial_id
+        trial_run_id
+        total_batches
+        end_time
+        metrics
+        batch_metrics
+    """
+
+    trial_id: int
+    trial_run_id: int
+    total_batches: int
+    end_time: datetime.datetime
+    metrics: Dict[str, Any]
+    group: str
+    batch_metrics: Optional[List[Dict[str, Any]]] = None
+
+    @classmethod
+    def _from_bindings(cls, metric_report: bindings.v1MetricsReport, group: str) -> "TrialMetrics":
+        key = "validation_metrics" if group == util._LEGACY_VALIDATION else "avg_metrics"
+        return cls(
+            trial_id=metric_report.trialId,
+            trial_run_id=metric_report.trialRunId,
+            total_batches=metric_report.totalBatches,
+            end_time=util.parse_protobuf_timestamp(metric_report.endTime),
+            metrics=metric_report.metrics[key],
+            batch_metrics=metric_report.metrics.get("batch_metrics", None),
+            group=metric_report.group,
+        )
+
+    @property
+    def steps_completed(self) -> int:
+        """@deprecated: Use total_batches instead."""
```

Comment:
why are we deprecating this?

Part of the reason we called `steps_completed` "steps completed" was that it wasn't actually required for it to be reported in terms of batches.  Has that changed?

Reply:
Also see the history for how we named [steps completed](https://hpe-aiatscale.slack.com/archives/CSLAGUF3M/p1651533616784609?thread_ts=1651520379.716279&cid=CSLAGUF3M)

Reply:
I actually don't know. This was from @hamidzr 's PR when we first committed the generic metrics: https://github.com/determined-ai/determined/pull/7407/files#diff-be8deb8e14b7fe99fa232a6339affeddb1e3a5dcf06990e967011b1cf2dbbeceR89

I merely refactored it here to a new module so that the metrics could be accessed from trials, checkpoints and model versions without causing cyclical dependencies.

Reply:
I was under the impression that this is an artifact of the old steps concept and that we wanted to get away from it since today the backing db schema is also called total_batches. I have no other reason for it.

Reply:
A reasonable misunderstanding, since we did used to have "steps" (circa 0.12.x) and those got removed fully in 0.16.0 I think, at which point we switched to "total_batches", but when we shipped Core API (0.18.0 I think) we pick a purposely-ambiguous name "steps_completed" for this concept because the master no longer cared what the unit was, and we thought the name in the public API should reflect that.

It stayed as `total_batches` in the database since it was conceptually the same value (no db migration required).

But yeah, this needs fixing.  This deprecation is not accurate, and we're exposing a legacy name from the database as the real name.  But we still require `steps_completed` as checkpoint metadata [here](https://github.com/determined-ai/determined/blob/5c839013e43f06365a4a5a0af1a37c8d99f588de/harness/determined/core/_checkpoint.py#L473), so only a few of the instances of `steps_completed` even got deprecated.

Reply:
I flipped the deprecations wherever I found `total_batches` in the python code, since this level of deprecation only really applies there. It doesn't seem that we are using it much within our own code. But I fixed all the usages we have within our tests and code.

I also deprecated the functions for `stream_<group>_metrics` in favor of the more general form that has the user specify the group name.